### PR TITLE
Fix issue 20734 - Allocate array literals as arguments to scope slice parameters on the stack

### PR DESCRIPTION
**Pull request by kinke**
Requires https://github.com/dlang/druntime/pull/3032 (specifically, adding `scope` to `__ArrayDtor()`). I've given this a quick test ride with LDC on Win64; druntime and Phobos unittests still pass; most `-checkaction=context` compile errors for the Phobos unittests are gone too.

### CODE REVIEW

**Comment by Geod24** [comment]
Nice! From the code, it doesn't seem to support multi-dimensional arrays though. Is that something we want ?

**Comment by kinke** [comment]
It works if the nested arrays are static, e.g., this:
```
void takeScopeSlice(const scope S[2][] slice) nothrow @nogc {}
...
takeScopeSlice([ [S(1), S(2)], [S(3), S(4)] ]);
assert(numDtor == 4);
```

Nested slices (`S[][]`) still require GC allocations though; that could be implemented as a further extension if somebody really wants that.

**Comment by MoonlightSentinel** [comment]
Why only for parameters? What about

```
scope arr = [ S(1), S(2) ];
takeScopeSlice(arr);
```

**Comment by kinke** [comment]
Ah, you guys.. ;) - progress is achieved incrementally, the aim here isn't to solve all existing problems at once.

`scope arr = [ S(1), S(2) ];` sounds interesting, but it's a) something different (would need to introduce an additional hidden static array variable in the same scope, not a temporary bound to the statement), and b) currently easy to work around by making it an explicit static array (`S[2] array = ...` - yes, the need to know the length in advance is a common annoyance) and then using it as slice.

**Comment by Geod24** [comment]
Agreed, while I would love to see `scope` array (but more `scope array = new int[42]` than literals), this PR is only focused on literals passed to functions.

**Comment by MoonlightSentinel** [comment]
Thanks for the explaination.

**Comment by kinke** [comment]
[Encountered DMD-specific `-betterC` bug for Windows MSVC targets filed as https://issues.dlang.org/show_bug.cgi?id=20737.]

**Review by Geod24 — APPROVED** [review]

**Comment by Geod24** [comment]
Thanks @kinke . Do you have plans for a new LDC release ? I would be very interested to make sure that `-checkaction=context` works for 1.21.0.

**Comment by kinke** [comment]
LDC 1.21 final will be released shortly after 2.091.1 and so include the fix wrt. temporaries, but not this breaking change. I guess you mean making sure the Phobos unittests can be compiled with `-checkaction=context`, that will still need a bit of love for the remaining issues (like handling `assert("myEnvVar" in environment)`, where `std.process.environment` is a class with static `opBinaryRight(string op : "in")`, which cannot be passed as function argument...).

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -2027,7 +2027,20 @@ private bool functionParameters(const ref Loc loc, Scope* sc,
...
                 if (a.op == TOK.cast_)
                     a = (cast(CastExp)a).e1;
-                if (a.op == TOK.function_)
+
+                ArrayLiteralExp ale;
+                if (p.type.toBasetype().ty == Tarray && !(p.storageClass & STC.return_) &&
+                    (ale = a.isArrayLiteralExp()) !is null)
+                {
+                    // allocate the array literal as temporary static array on the stack
+                    ale.type = ale.type.nextOf().sarrayOf(ale.elements.length);
```
Comment (WalterBright) [inline src/dmd/expressionsem.d:2036]: `ale.elements` could be `null` causing a seg fault

**Comment by Geod24** [comment]
Just found out there was a previously existing issue for this: https://issues.dlang.org/show_bug.cgi?id=11657
Martin also opened https://issues.dlang.org/show_bug.cgi?id=11658 around the same time.

File: src/dmd/expressionsem.d (hunk 1 above)
Comment (kinke) [reply src/dmd/expressionsem.d:2036]: Thx, fixed in #11126.

**Pull request merged**